crates/aptos-rosetta/src/types/objects.rs

```diff
@@ -497,187 +500,61 @@ impl Display for TransactionType {
 }
 
 impl Transaction {
-    pub async fn from_transaction(
-        coin_cache: Arc<CoinCache>,
-        rest_client: &aptos_rest_client::Client,
-        txn: aptos_rest_client::Transaction,
-    ) -> ApiResult<Transaction> {
+    pub async fn from_transaction(txn: aptos_rest_client::Transaction) -> ApiResult<Transaction> {
         use aptos_rest_client::Transaction::*;
-        let (txn_type, maybe_sender, txn_info, events) = match txn {
+        let (txn_type, maybe_sender, txn_info, events, maybe_payload) = match txn {
             // Pending transactions aren't supported by Rosetta (for now)
             PendingTransaction(_) => return Err(ApiError::TransactionIsPending),
             UserTransaction(txn) => (
                 TransactionType::User,
                 Some(txn.request.sender),
                 txn.info,
                 txn.events,
+                Some(txn.request.payload),
+            ),
+            GenesisTransaction(txn) => (TransactionType::Genesis, None, txn.info, txn.events, None),
+            BlockMetadataTransaction(txn) => (
+                TransactionType::BlockMetadata,
+                None,
+                txn.info,
+                txn.events,
+                None,
+            ),
+            StateCheckpointTransaction(txn) => (
+                TransactionType::StateCheckpoint,
+                None,
+                txn.info,
+                vec![],
```

Comment:
I'm sorta surprised this doesn't emit events.

Reply:
The state checkpoint transaction doesn't actually commit any state changes, which is why it has no events.